### PR TITLE
Metric repository perf

### LINKED ISSUE #1900: Non-optimal app layer iteration in MetricRepository
Before submitting your issue, please make sure that you've checked all of the checkboxes below.
- [X] You're running the [latest release](https://github.com/CachetHQ/Cachet/releases/latest) version of Cachet.
- [X] Ensure that you're running at least PHP 5.5.9, you can check this by running `php -v`
- [X] You've ran `rm -rf bootstrap/cache/*` from the root of your Cachet installation.

To help us better understand your issue, please answer the following — cheers!
### Your setup
- _What version of Cachet?_ v2.3.0-RC5
- _What database driver? MySQL?_ 5.5.47
- _What version of PHP?_ 5.6
### Expected behaviour

Iterating over the metric points should be quite performant.
### Actual behaviour

I've noticed in the MetricRepository you are iterating over each minute/hour in the app layer - which will potentially generate between 12 - 60 queries of an average return time of 4-8 seconds each - depending on the data size. 

Subsequently, we're seeing our MySql process jump to 100% CPU time once the Metric Points table gets close to 40,000 records

Wouldn't it be more performant to iterate this in the query itself? ie.MySql example

`
SELECT AVG(mp.value * mp.counter) AS value
FROM metrics m
  INNER JOIN metric_points mp ON m.id = mp.metric_id
WHERE m.id = 1 AND DATE_SUB(curdate(), INTERVAL 12 HOUR) <= mp.created_at
GROUP BY HOUR(mp.created_at)
`

### PR DESCRIPTION
**Pull request by jbrooksuk**
Closes #1900.

---

Refactoring metrics. With this, I intend to fix the performance of metrics across all three database drivers.

Note that this is probably going to end up bigger than I intended it to be, but it'll make it easier to deal with.

Todo:

- [x] Refactor all metric repositories to query all metric points between a given time span.
- [x] Stop using static table names as the tables may be prefixed.

### CODE REVIEW

File: app/Foundation/Providers/RepositoryServiceProvider.php [hunk 1]
```diff
@@ -13,9 +13,9 @@
 
 use CachetHQ\Cachet\Dates\DateFactory;
 use CachetHQ\Cachet\Repositories\Metric\MetricRepository;
-use CachetHQ\Cachet\Repositories\Metric\MySqlRepository as MetricMySqlRepository;
-use CachetHQ\Cachet\Repositories\Metric\PgSqlRepository as MetricPgSqlRepository;
-use CachetHQ\Cachet\Repositories\Metric\SqliteRepository as MetricSqliteRepository;
+use CachetHQ\Cachet\Repositories\Metric\MySql as MetricMySqlRepository;
```
Comment (GrahamCampbell) [inline app/Foundation/Providers/RepositoryServiceProvider.php]: I don't like these renames. ;(

Comment (jbrooksuk) [inline app/Foundation/Providers/RepositoryServiceProvider.php]: What would you suggest?

Comment (GrahamCampbell) [inline app/Foundation/Providers/RepositoryServiceProvider.php]: Keeping it how it was. :P

Comment (jbrooksuk) [inline app/Foundation/Providers/RepositoryServiceProvider.php]: Hah, okay then. I'll rename them once it's finished.

File: app/Repositories/Metric/MySql.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+     *
+     * @param \CachetHQ\Cachet\Models\Metric $metric
+     * @param int                            $minutes
+     *
+     * @return \Illuminate\Support\Collection
+     */
+    public function getPointsSinceMinutes(Metric $metric, $minutes)
+    {
+        $queryType = $this->getQueryType($metric);
+        $points = Collection::make(DB::select("SELECT DATE_FORMAT(metric_points.`created_at`, '%H:%i') AS `key`, {$queryType} FROM {$this->getTableName()} INNER JOIN metric_points ON metrics.id = metric_points.metric_id WHERE metrics.id = :metricId AND metric_points.`created_at` >= DATE_SUB(NOW(), INTERVAL :minutes MINUTE) GROUP BY HOUR(metric_points.`created_at`), MINUTE(metric_points.`created_at`) ORDER BY metric_points.`created_at`", [
```
Comment (ConnorVG) [inline app/Repositories/Metric/MySql.php]: These are absolutely disgusting... :scream:

Comment (GrahamCampbell) [inline app/Repositories/Metric/MySql.php]: :trollface:

Comment (jbrooksuk) [inline app/Repositories/Metric/MySql.php]: Laravel 5.3 stops us having to wrap `DB::select` in a collection at least.

**Comment by jbrooksuk** [comment]
The issue we have with all of the metrics is that each DB driver queries differently and Eloquent isn't flexible enough to allow us to keep the same query structure — unless I'm missing something?

I really want to get this into 2.4, so I'll try and finish this at a later date.

**Comment by jbrooksuk** [comment]
Rebased.

**Comment by jbrooksuk** [comment]
This is finally done!

**Comment by ConnorVG** [comment]
😱 👍

**Pull request merged**